Support frondend in E2E transformer

kamo-naoyuki opened the pull request:
I modified E2E transformer to support frontend.

- [x] Add tests
- [x] Implement FrondendInterface and support dynamic importing scheme for fronend.

sw005320 commented:
Can you add a test script?

ShigekiKarita commented on espnet/nets/pytorch_backend/e2e_asr_transformer.py:
```diff
@@ -224,7 +249,14 @@ def recognize(self, feat, recog_args, char_list=None, rnnlm=None, use_jit=False)
-        feat = torch.as_tensor(feat).unsqueeze(0)
+        feat = to_torch_tensor(feat).unsqueeze(0)
+        # 0. Frontend
+        if self.frontend is not None:
```
I do not like this. Remember our `recognize` method is already too complex.

How about this?
```python

# espnet.nets.pytorch_backend.frontend_asr.py
class FrontendASR(torch.nn.Module):
    def __init__(self, frontend, asr_model):
        super().__init__()
        assert isinstance(asr_model, ASRInterface)
        self.frontend = frontend
        self.asr_model = asr_model

    def forward(self, xs_pad, ilens, ys_pad):
        hs_pad, hlens, _ = self.frontend(to_torch_tensor(xs_pad), ilens)
        hs_pad, hlens = self.feature_transform(hs_pad, hlens)
        return self.asr_model(hs_pad, hlens, ys_pad)

    def recognize(self, feat, recog_args, char_list=None, rnnlm=None, **kwargs):
        self.eval()
        feat = to_torch_tensor(feat).unsqueeze(0)
        feat, _, _ = self.frontend(feat, [feat.size(1)])
        feat, _ = self.feature_transform(feat, [feat.size(1)])
        feat = feat.squeeze(0)
        return self.asr_model(feat, recog_args, char_list, rnnlm, **kwargs)

    def enfance(...):
        ...
```

ShigekiKarita replied:
If `self.frontend.forward` applies `to_torch_tensor` inside itself, we can generalize this class between pytorch and chainer.

kamo-naoyuki replied:
I'm not sure what you propose. 
Does FrontendAsr correspond to the current `E2E`? I can't get  how to instantiate the class. Do you mean this?

```python
model = FrontendASR(FrondEnd(...), OriginalE2E(...))
```

ShigekiKarita replied:
maybe you can add `--frontend-module espnet.nets.xxx:YYY` and asr.py can be changed like
``` python
model = dynamic_import(args.model_module)
...
if args.frontend_module is not None:
    frontend = dynamic_import(args.frontend_module)
    model = FrontendASR(frontend, model)
```

kamo-naoyuki replied:
O.K. @ShigekiKarita 's plan looks nice. 
I was also concerned about modularity of FrontEnd. 

I'll notify if I finished.

kamo-naoyuki commented:
> Can you add a test script?

OK

kamo-naoyuki commented:
Finished. 
 I put out the fronted feature from E2E class and I implemented `FrontendASRInterface` and `FrontendASR`. Now we acquire two features.

- Frontend module can be used for arbitrary E2E class, including E2E for Transformer. 
- FrontendASR itself can be replaced by using `--frontend-module`. 

@Jzmo I changed `--use-frontend` to `--frontend-module`. Please reflect after merged.

ShigekiKarita commented on espnet/nets/pytorch_backend/frontend_asr.py:
```diff
@@ -0,0 +1,212 @@
+    def add_arguments(parser: Union[argparse.ArgumentParser,
+                                    configargparse.ArgumentParser]):
+        # WPE related
+        parser.add_argument('--use-wpe', type=strtobool, default=False,
```
use group. see transformer

ShigekiKarita commented on espnet/asr/pytorch_backend/asr.py:
```diff
@@ -609,11 +629,23 @@ def enhance(args):
     model.recog_args = args
 
+    if train_args.frontend_module is None:
+        raise RuntimeError('--frontend_module is not set when training.')
```
--frontend-module

ShigekiKarita commented on espnet/asr/pytorch_backend/asr.py:
```diff
@@ -609,11 +629,23 @@ def enhance(args):
-    assert isinstance(model, ASRInterface)
-    torch_load(args.model, model)
+    if train_args.frontend_module is not None:
+        # FIXME(kamo): Assuming fbank feature, so you need to change with the other frontend type
```
remove this if block. FrontendASR can set its own `idim` by `train_args`

kamo-naoyuki replied:
I can't understand out what you mean. This block is inevitable.

ShigekiKarita commented on espnet/asr/pytorch_backend/asr.py:
```diff
@@ -247,9 +244,15 @@ def train(args):
-    assert isinstance(model, ASRInterface)
+    if args.frontend_module is not None:
+        # FIXME(kamo): Assuming fbank feature, so you need to change with the other frontend type
+        _idim = args.n_mels
```
ditto

ShigekiKarita commented on espnet/nets/pytorch_backend/frontend_asr.py:
```diff
@@ -0,0 +1,212 @@
+        :param torch.Tensor xs_pad: batch of padded input sequences (B, Tmax, idim)
+        :param torch.Tensor ilens: batch of lengths of input sequences (B)
+        :param torch.Tensor ys_pad: batch of padded character id sequence tensor (B, Lmax)
+        :return: ctc loass value
```
loss

ShigekiKarita replied:
I think ASRInterface returns single loss value now. see https://github.com/espnet/espnet/blob/master/espnet/nets/asr_interface.py#L20

ShigekiKarita commented on espnet/nets/pytorch_backend/frontend_asr.py:
```diff
@@ -0,0 +1,212 @@
+                            help='The sample frequency used for '
+                                 'the mel-fbank creation.')
+        parser.add_argument('--n-mels', type=int, default=80,
+                            help='The number of mel-frequency bins.')
```
Mel

ShigekiKarita replied:
considering `--fbank-xxx` options, `--fbank-dim` is more clear

kamo-naoyuki replied:
Mel?

ShigekiKarita commented on espnet/nets/pytorch_backend/frontend_asr.py:
```diff
@@ -0,0 +1,212 @@
+        # Beamformer related
+        parser.add_argument('--use-beamformer', type=strtobool,
+                            default=True, help='')
+        parser.add_argument('--btype', default='blstmp', type=str,
```
Can you make these `--bxxx` more expressive like `--masknet-type` or something? because we have already added breaking change in the option `--use-frontend`

ShigekiKarita commented:
@sw005320 Because of the breaking change `--use-frontend` to `--frontend-module`, should we merge this PR into v0.5.0?

ShigekiKarita commented:
Nice generalization and coverage increase.

This PR looks great except for my small comments above. Some comments might require more discussion. Through this PR, I feel `XXXInterface` is really important to build a nice framework (maybe `ScorerInterface` for beam search is my next work).

sw005320 commented:
> @sw005320 Because of the breaking change --use-frontend to --frontend-module, should we merge this PR into v0.5.0?

I agree.
I'll change it.

Shinji

kamo-naoyuki commented:
I mistakenly closed. I reopened at  #890
